### DISCUSSION TITLE
How to convert io.trino.jdbc.Row to java object

### DISCUSSION THREAD

**Discussion by ceciliazcx** [trinodb/trino#14923] [Q&A]
The data set I work with is nested data. `resultSet.getObject(id)`  will  return the data as following (but much more complicated) and data type is `io.trino.jdbc.Row`.
```{
field1={name=name1}, 
field2=[{id=id1, value=value1}, {id=id2, value=value2}]
}
```
I tried to read data from`io.trino.jdbc.Row` but `getFields()` does not work. and it seems the only way I can do is `toString()`. Is there other thing I can try to read from `io.trino.jdbc.Row`?

**Comment by hashhar** [comment]
`getFields` seems to work for me - what issue are you running into when you say "but `getFields()` does not work"?

The other alternative is to read into a `Map` by doing `getObject(column, Map.class)`.

**Reply by ceciliazcx** [reply]
Thanks for your replay. I suppose I did not make my question clear. Sorry about it. let me make more details. here are two senarios:
1. `res.getObject(listOfMap)` will return `trinoArray` and I need `List<Map<String, String>>`. I tried `(List<Map<String, String>>) res.getObject(columnName, List.class));` but got errors. Is there any other way I can convert it?
2. `res.getObject(customObject)` will return `trinoRow` and the actual type is `CustomObject`. I think I can do `trinoRow.getFields().get(1).getValue().get(0).getFields().get(0).getValue()...`. Then create a new  `CustomObject` and put value. I wonder if there is a better we can handle it? Thanks a lot for your help.

**Reply by hashhar** [reply]
So you want to map rows into objects - this is generally not something the JDBC driver is responsible for. You'll either need to create a function that takes a ResultSet as input and outputs a custom object created using that result set.
You can use libraries like Jdbi or some ORM to help with the mapping part.

Regarding 1 - here's some code in the form of test code which shows how JDBC and arrays work - this is true for all JDBC drivers - not something specific to Trino.

```java
ResultSet rs = statement.query("VALUES ARRAY[map_from_entries(ARRAY[('k1', 42), ('k2', null)]), map_from_entries(ARRAY[('l1', 42), ('l2', null)])];");

// Java representation of what we expect to get back
Map<String, Integer> expectedElementOne = new HashMap<>();
expectedElementOne.put("k1", 42);
expectedElementOne.put("k2", null);

Map<String, Integer> expectedElementTwo = new HashMap<>();
expectedElementTwo.put("l1", 42);
expectedElementTwo.put("l2", null);

// Let's read using JDBC
// Option 1
Array array = rs.getArray(1);
assertThat(array.getArray()).isEqualTo(new Object[] {expectedElementOne, expectedElementTwo});

// Option 2
array = (Array) rs.getObject(1);
assertThat(array.getArray()).isEqualTo(new Object[] {expectedElementOne, expectedElementTwo});

// Option 3
assertEquals(rs.getObject(1, List.class), List.of(expectedElementOne, expectedElementTwo));
```
